// sqreen/lib/backend/login.js
/**
 * Copyright (c) 2016 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Logger = require('../logger');
const Path = require('path');
const Os = require('os');

const Reader = require('../package-reader');
const Version = require('../../version.json').version;

let appPackage = null;

module.exports.getPkg = function () {

    return appPackage;
};

/**
 * internal method
 * build login payload
 */
const buildPayload = module.exports._buildPayload = function (deps_signature) {

    const pkg = Reader.readPackage(Path.join(process.cwd(), 'package.json'));
    Logger.DEBUG('is building the login payload');

    appPackage = pkg;

    return {
        bundle_signature: deps_signature,
        various_infos: {
            time: new Date(),
            app_name: pkg && pkg.name || null,
            app_version: pkg && pkg.version || null,
            // declared_dependencies: pkg && pkg.dependencies || null,
            // declared_devdependencies: pkg && pkg.devDependencies || null,
            pid: process.pid,
            ppid: null,
            euid: process.geteuid && process.geteuid(),
            egid: process.getegid && process.getegid(),
            uid: process.getuid && process.getuid(),
            gid: process.getgid && process.getgid(),
            name: process.title
        },
        agent_type: 'nodejs',
        agent_version: Version,
        os_type: Os.arch() + '-' + Os.type(),
        hostname: Os.hostname(),
        runtime_type: 'node',
        runtime_version: process.version,
        framework_type: '',
        framework_version: null,
        environment: process.env.NODE_ENV
    };
};

module.exports.getPayload = function () {

    return new Promise((resolve) => {

        Reader.getDependenciesHash((sig) => {

            return resolve(buildPayload(sig));
        });
    });
};
