// sqreen/lib/config/index.js
/**
 * Copyright (c) 2016 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
/**
 * This script is to be run at startup, therefore synchronous operations are permitted
 */
'use strict';
const Logger = require('../logger');
const Joi = require('joi');
const Fs = require('fs');
const Path = require('path');

const DEFAULT = require('./default');

const LOG_LEVELS = Object.keys(Logger.levels);

const configSchema = Joi.object().keys({
    url: Joi.string().uri().default(DEFAULT.url),
    token: Joi.string().required(),
    local_rules: Joi.string().optional(),
    rules_verify_signature: Joi.boolean().truthy('true', 'yes', 1, '1').falsy('false', 'no', 0, '0').default(DEFAULT.rules_verify_signature),
    log_level: Joi.string().only(LOG_LEVELS).default(DEFAULT.log_level),
    log_location: Joi.string().default(DEFAULT.log_location),
    run_in_test: Joi.boolean().truthy('true', 'yes', 1, '1').falsy('false', 'no', 0, '0').default(DEFAULT.run_in_test),
    block_all_rules: Joi.boolean().truthy('true', 'yes', 1, '1').falsy('false', 'no', 0, '0').default(DEFAULT.block_all_rules),
    report_perf_newrelic: Joi.boolean().truthy('true', 'yes', 1, '1').falsy('false', 'no', 0, '0').default(DEFAULT.report_perf_newrelic),
    initial_features: Joi.string().default(DEFAULT.initial_features)
});

/**
 * config validator
 * @param rawConfig
 */
const parseConfig = function (rawConfig) {

    if (rawConfig.log_level && LOG_LEVELS.lastIndexOf(rawConfig.log_level) < 0){
        rawConfig.log_level = Logger.logLevels.WARN;
    }

    const result = Joi.validate(rawConfig, configSchema);

    if (result.error) {
        // if an error happens, it is logged and return nothing
        Logger.ERROR(result.error.message);
        return;
    }
    const config = result.value;
    Logger.transports.console.level = config.log_level;

    Logger.addFileTransport(config.log_location, config.log_level);

    if (config.local_rules) {
        try {
            const Rules = require('../rules'); // load here and not at the beginning of the script to prevent circular import issue
            Rules.enforceRuleList(JSON.parse(Fs.readFileSync(config.local_rules, 'utf-8')), !config.rules_verify_signature);
        }
        catch (err) {
            Logger.DEBUG(`Could not load local rules ${err}`);
        }
    }

    if (config.initial_features) {
        try {
            require('../command/features').change(JSON.parse(Fs.readFileSync(config.initial_features, 'utf-8')));
        }
        catch (err) {
            Logger.DEBUG(`Could not load initial features ${err}`);
        }
    }

    return config;
};

const asBoolean = function (str) {

    return str === 'true' || str === '1';
};

/**
 * read conf, first in a file then oevrrides it with env variables
 */
const readConfigFile = function (path) {

    try {
        const readJson = Fs.readFileSync(path, 'utf-8');
        return JSON.parse(readJson);
    }
    catch (err) {
        Logger.INFO(`Sqreen could not parse content of ${path}. Is it a valid JSON file ?`);
    }
};

const readConfig = function () {

    let config = {};

    // get the base dir of the process
    let path;
    if (process.env.SQREEN_CONFIG_FILE) {
        path = process.env.SQREEN_CONFIG_FILE;
    }
    else {
        const baseDir = process.cwd();

        // get the content of this directory
        const baseDirContent = Fs.readdirSync(baseDir);

        // if a configuration exists in a json file, we use it
        if (baseDirContent.indexOf('sqreen.json') > -1) {
            // we checked that the file existed
            path = Path.join(baseDir, 'sqreen.json');
        }
    }

    if (path) {
        config = readConfigFile(path) || config;
    }

    if (config.disable) {
        config.run_in_test = config.disable;
        delete config.disable;
    }

    // env variables overrides file content.
    if (process.env.SQREEN_URL) {
        config.url = process.env.SQREEN_URL;
    }
    if (process.env.SQREEN_TOKEN) {
        config.token = process.env.SQREEN_TOKEN;
    }
    if (process.env.SQREEN_RULES) {
        config.local_rules = process.env.SQREEN_RULES;
    }
    if (process.env.SQREEN_RULES_SIGNATURE) {
        config.rules_verify_signature = asBoolean(process.env.SQREEN_RULES_SIGNATURE);
    }
    if (process.env.SQREEN_LOG_LEVEL) {
        config.log_level = process.env.SQREEN_LOG_LEVEL;
    }
    if (process.env.SQREEN_LOG_LOCATION) {
        config.log_location = process.env.SQREEN_LOG_LOCATION;
    }
    if (process.env.SQREEN_RUN_IN_TEST) {
        config.run_in_test = asBoolean(process.env.SQREEN_RUN_IN_TEST);
    }
    if (process.env.SQREEN_DISABLE) {
        config.run_in_test = asBoolean(process.env.SQREEN_DISABLE);
    }
    if (process.env.SQREEN_BLOCK_ALL_RULES) {
        config.block_all_rules = asBoolean(process.env.SQREEN_BLOCK_ALL_RULES);
    }
    if (process.env.SQREEN_REPORT_PERF_NR) {
        config.report_perf_newrelic = asBoolean(process.env.SQREEN_REPORT_PERF_NR);
    }
    if (process.env.SQREEN_INITIAL_FEATURES) {
        config.initial_features = process.env.SQREEN_INITIAL_FEATURES;
    }

    return config;
};

let agentConfig;
/**
 * load and cache the config
 * @param {boolean} [force] force the reloading of th config.. use with caution
 * @returns {*}
 */
const getConfig = function (force) {

    if (!agentConfig || force) {
        agentConfig = parseConfig(readConfig());
    }
    return agentConfig;
};


module.exports = {
    parseConfig,
    readConfig,
    getConfig
};
