// sqreen/lib/rules/rules-callback/userAgentRegexp.js
/**
 * Copyright (c) 2016 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const RegexpRule = require('./regexpRule');
const CB_STATUS = require('../../enums/cbReturn').STATUS;

const UARegexpCB = class {

    constructor(rule) {

        this.matcherList = rule.data.values.map((val) => RegexpRule.getRegexpMatcherObject(val));
        this.superMatcher = RegexpRule.getRegexpMatcherObject(rule.data.values.map((str) => str.replace('.*.', '').replace('.*', '')).join('|'));
        this.rule = rule;
        const self = this;
        this.pre = function (args) {

            if (args[0] === 'request' && args[1] && args[1].headers) {
                return self.executePre(args[1].headers['user-agent'], args);
            }
            return null;
        };
    }

    executePre(userAgent, args) {

        if (!userAgent) {
            return null;
        }

        if (!this.superMatcher.match(userAgent)) {
            return null;
        }

        for (let i = 0; i < this.matcherList.length; ++i) {

            if (this.matcherList[i].match(userAgent)) {

                const result = { record: { found: this.matcherList[i].pattern }, originalSession: { req: args[1], res: args[2] } };
                if (this.rule.block) {
                    result.status = CB_STATUS.RAISE;
                }
                return result;
            }
        }
        return null;
    }


};
module.exports.getCbs = function (rule_) {

    if (!rule_ || !rule_.data || !rule_.data.values) {
        //noinspection JSValidateTypes
        return null;
    }

    return new UARegexpCB(rule_);
};
