// sqreen/lib/instrumentation/reportUtil.js
/**
 * Copyright (c) 2016 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
module.exports.mapRequest = function (req) {

    req.headers = req.headers || {};
    req.connection = req.connection || {};

    return {
        rid: req.__sqreen_uuid,
        user_agent: req.headers['user-agent'],
        scheme: 'http',
        verb: req.method,
        host: req.headers.host,
        port: req.headers.host && req.headers.host.split(':').pop(),
        rport: req.connection.remotePort,
        path: req.url,
        addr: req.headers['x-forwarded-for'] || req.connection.remoteAddress,
        referer: req.headers.referer
    };
};

const RuleUtils = require('../rules/rules-callback/utils');
module.exports.mapRequestParams = function (req) {

    const claims = RuleUtils.getLookableClaims(req);

    if (claims.__sqreen_lookup && claims.__sqreen_lookup.hapi) {
        return {
            scheme: req.scheme,
            query: claims.__sqreen_lookup.hapi.query,
            body: claims.__sqreen_lookup.hapi.payload
        };
    }
    return {
        scheme: req.scheme,
        body: req.body,
        query: req.query
    };
};
